src/zkevm_specs/evm/execution/sar.py

```diff
@@ -0,0 +1,172 @@
+        stack_pointer=Transition.delta(1),
+    )
+
+
+def __check_witness(
+    instruction: Instruction,
+    a: RLC,
+    shift: RLC,
+    b: RLC,
+    a64s: Sequence[FQ],
+    b64s: Sequence[FQ],
+    a64s_lo: Sequence[FQ],
+    a64s_hi: Sequence[FQ],
+    shf_div64,
+    shf_mod64,
+    p_lo,
+    p_hi,
+    p_top,
+    is_neg,
+):
+    shf_lt256 = instruction.is_zero(instruction.sum(shift.le_bytes[1:]))
+    for idx in range(4):
+        offset = idx * N_BYTES_U64
+
+        # a64s constraint
+        instruction.constrain_equal(
+            a64s[idx],
+            instruction.bytes_to_fq(a.le_bytes[offset : offset + N_BYTES_U64]),
+        )
+
+        # b64s constraint
+        instruction.constrain_equal(
+            b64s[idx] * shf_lt256 + is_neg * (1 - shf_lt256) * MAX_U64,
+            instruction.bytes_to_fq(b.le_bytes[offset : offset + N_BYTES_U64]),
+        )
+
+        # `a64s[idx] == a64s_lo[idx] + a64s_hi[idx] * p_lo`
+        instruction.constrain_equal(a64s[idx], a64s_lo[idx] + a64s_hi[idx] * p_lo)
+
+        # `a64s_lo[idx] < p_lo`
+        a64s_lo_lt_p_lo, _ = instruction.compare(a64s_lo[idx], p_lo, N_BYTES_U64)
+        instruction.constrain_equal(a64s_lo_lt_p_lo, FQ(1))
+
+    # merge contraints
+    shf_div64_eq0 = instruction.is_zero(shf_div64)
+    shf_div64_eq1 = instruction.is_zero(shf_div64 - 1)
+    shf_div64_eq2 = instruction.is_zero(shf_div64 - 2)
+    shf_div64_eq3 = instruction.is_zero(shf_div64 - 3)
+    instruction.constrain_equal(
+        b64s[0],
+        (a64s_hi[0] + a64s_lo[1] * p_hi) * shf_div64_eq0
+        + (a64s_hi[1] + a64s_lo[2] * p_hi) * shf_div64_eq1
+        + (a64s_hi[2] + a64s_lo[3] * p_hi) * shf_div64_eq2
+        + (a64s_hi[3] + p_top) * shf_div64_eq3
+        + is_neg * MAX_U64 * (1 - shf_div64_eq0 - shf_div64_eq1 - shf_div64_eq2 - shf_div64_eq3),
+    )
+    instruction.constrain_equal(
+        b64s[1],
+        (a64s_hi[1] + a64s_lo[2] * p_hi) * shf_div64_eq0
+        + (a64s_hi[2] + a64s_lo[3] * p_hi) * shf_div64_eq1
+        + (a64s_hi[3] + p_top) * shf_div64_eq2
+        + is_neg * MAX_U64 * (1 - shf_div64_eq0 - shf_div64_eq1 - shf_div64_eq2),
+    )
+    instruction.constrain_equal(
+        b64s[2],
+        (a64s_hi[2] + a64s_lo[3] * p_hi) * shf_div64_eq0
+        + (a64s_hi[3] + p_top) * shf_div64_eq1
+        + is_neg * MAX_U64 * (1 - shf_div64_eq0 - shf_div64_eq1),
+    )
+    instruction.constrain_equal(
+        b64s[3],
+        (a64s_hi[3] + p_top) * shf_div64_eq0 + is_neg * MAX_U64 * (1 - shf_div64_eq0),
+    )
+
+    # shift constraint
+    instruction.constrain_equal(
+        instruction.bytes_to_fq(shift.le_bytes[:1]),
+        shf_mod64 + shf_div64 * 64,
+    )
+
+    # `p_lo == pow(2, shf_mod64)` and `p_hi == pow(2, 64 - shf_mod64)`.
+    instruction.pow2_lookup(shf_mod64, p_lo, FQ(0))
+    instruction.pow2_lookup(64 - shf_mod64, p_hi, FQ(0))
+
+
+def __gen_witness(instruction: Instruction, opcode: FQ, a: RLC, shift: RLC):
```

Comment:
If this is going to be common for `SAR` and `SHL`  we should probably have it in a sepparated file.

Reply:
Given that [SHL & SHR](https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/578/files#diff-a841e8152bd12f5377160e6127ef1f622161f8bf52df6c62574ce7c0f7c643de) are implemented in a new way, I think this case only applies to SAR